shared/src/main/scala/io/kaitai/struct/translators/GoTranslator.scala

```
@@ -253,6 +253,11 @@ class GoTranslator(out: StringLanguageOutputWriter, provider: TypeProvider, impo
     }
   }
 
+  override def doRegexMatchOp(str: String, regex: String): String = {
+    importList.add("regexp")
+    s"matched, _ := regexp.MatchString(`${regex}`, ${trStringLiteral(str)}); matched"
```

Will this fail if it will be used twice in the same function?

---

And I think it's not wise to ignore the error in case it occurs (doing `matched, _`). Please follow the same approach used in the rest of the `GoTranslator`, notice that the methods you added are ones of the few in the file returning `String` and not `TranslatorResult`. There is as special method `outVarCheckRes` handling the errors returned from functions, I suggest using that one.

---

I tried to modify the translateExpr func in GoTranslator, but it seems to me that this function is not executed, am I wrong ?

---

> Will this fail if it will be used twice in the same function?

Yes, it probably will, because the variables can't be redeclared in the same scope in Go [unless some new variables are introduced](https://yourbasic.org/golang/redeclaring-variables/). I've made a simple test: https://repl.it/repls/WonderfulJointProcedure

> I tried to modify the translateExpr func in GoTranslator, but it seems to me that this function is not executed

@jocelynke Go language didn't play well with the classic `BaseTranslator` model that returns `String`s, because of the error handling in Go. Go doesn't have `try/catch` as other languages, it treats errors as values. This means that if you want the error from a function you call to bubble up in the call stack, it won't happen automatically. You have to check if the `error` value returned from the function is `nil` and if it's not, then an error occured and you should `return` it to the caller. (_Just returning it isn't exactly [debug-friendly](https://medium.com/rungo/error-handling-in-go-f0125de052f0#7fea), because the caller loses any context which function in the stack actually caused the error, but it's unfortunately done this way in KS currently._)

This means that parsing a single `u1` field from the stream can't be done in a single expression (because it can of course result in an EOF error), but instead it is a 3-phase process:

```go
func (this *HelloWorld) Read(...) (err error) {
 // ...
    tmp1, err := this._io.ReadU1()
    if err != nil {
        return err
    }
    this.One = tmp1
```

Because of this, `GoTranslator` can't just extend `BaseTranslator` as other languages do, because the methods in  `BaseTranslator` have return type `String` which doesn't fit for Go (and it can't be changed in the derived classes). Not all methods in `GoTranslator` can be implemented in a single expression which can be passed as `String`, for these the return value of the called error-returning method is saved to a temporary variable (named `tmp1`, `tmp2`, ... to avoid name conflicts), the error is handled and only the "serial number" of the temp variable is returned. This is done by the method [`outVarCheckRes`](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/shared/src/main/scala/io/kaitai/struct/translators/GoTranslator.scala#L449-L454).

Hence most methods in `GoTranslator` return `TranslatorResult`, which is defined on the top:

https://github.com/kaitai-io/kaitai_struct_compiler/blob/ef7178cb2e1dae8c443214e825bcf361e1e158ee/shared/src/main/scala/io/kaitai/struct/translators/GoTranslator.scala#L11-L13

And now how you should do it. I think the `doRegexMatchOp` here could look like this:

```scala
override def doRegexMatchOp(str: String, regex: String): TranslatorResult = {
  importList.add("regexp")
  outVarCheckRes(s"regexp.MatchString(`${regex}`, ${trStringLiteral(str)})")
```

Then you'll have to move the abstract definition of this method from `CommonOps` to `BaseTranslator`:

https://github.com/kaitai-io/kaitai_struct_compiler/blob/7120c8ec65c7b5d35dc5c17333dace47470010ce/shared/src/main/scala/io/kaitai/struct/translators/CommonOps.scala#L68

The reason is that `CommonOps` is inherited by `GoTranslator` too (whereas `BaseTranslator` is not), but in `GoTranslator` you need to specify a different return type, which won't work in Scala. And the `Ast.expr.RegexMatch` handling in `BaseTranslator`

https://github.com/kaitai-io/kaitai_struct_compiler/blob/7120c8ec65c7b5d35dc5c17333dace47470010ce/shared/src/main/scala/io/kaitai/struct/translators/BaseTranslator.scala#L105-L112

needs to be in `GoTranslator` as well, because again, it doesn't extend `BaseTranslator`.

---

Thank you for this deep explanation, I had implemented what you told me, but the thing is I have no log of a call to translateExpr in the GoTranslator. The consequence is that no code is generated to validate the field that needs it.

---

@generalmimon I found that the valid key is not supported by Go in kaitai, so it's normal that no code is generated...